cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,608 @@
+# Plugable Cloud Provider over gRPC
+
+
+Authors:
+
+* Hector Fernandez <<hfernandez@d2iq.com>> (@hectorj2f)
+
+
+## Motivation
+
+CA is released as a bundle which includes a hardcoded list of supported cloud providers.
+Whenever users want to implement the logic of their own cloud provider, they need to fork the CA and add their own implementation.
+
+In particular users need to follow these steps to support a custom private cloud:
+
+* Write a client for your private cloud in Go, implementing CloudProvider interface.
+
+* Add constructing it to cloud provider builder.
+
+* Build a custom image of Cluster Autoscaler that includes those changes and configure it to start with your cloud provider.
+
+This is a concern that has been raised in the past [PR953](https://github.com/kubernetes/autoscaler/issues/953) and [PR1060](https://github.com/kubernetes/autoscaler/issues/1060).
+
+Therefore a new implemetation should be added to CA in order to extend it without breaking any backwards compatibility or
+the current cloud provider implementations.
+
+## Goals
+
+* Support custom cloud provider implementations without changing the current `CloudProvider` interface.
+* Make CA extendable, so users do not need to fork the CA repository.
+
+## Proposal
+
+There are couple of examples of pluggable designs using Go SDKs that would guide us on how to extend CA
+to support custom providers as plugins.
+This approach is inspired based on [Hashicorp go-plugin](https://github.com/hashicorp/go-plugin) and [Grafana Go SDK for plugins](https://github.com/grafana/grafana-plugin-sdk-go).
+
+There are two alternatives on how to plug custom cloud providers:
+
+* **Option1:** Install the plugin as a binary that would be mounted into the CA container
+and invoked by CA server.
+In this option CA server launches each provider plugin as a subprocess and communicates with it over gRPC.
+
+* **Option2:** A custom cloud provider server is deployed along side CA and both communicates via gRPC with TLS/SSL.
```

Comment:
i just want to understand these options more thoroughly.

in option 1, would the CA be deploying the subprocess?
would it deployed in the same container?

in option 2, the plugin would be started as a separate container (eg from a deployment or similar)?

Reply:
Yeah, I left this open but I am more inclined to Option2.

Reply:
I prefer Option2 at least as part of an initial version. wdyt ?

Reply:
i also prefer option 2. i think option 1 does not follow what i would expect for kubernetes applications with respect to deploying/managing multiple processes in the same container.

additionally, it makes the architecture much clearer to a user who might need to customize how these containers are deployed.

Reply:
I totally agree. I'll change that too.

Reply:
Done